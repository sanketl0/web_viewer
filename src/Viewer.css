/*
 *  Sizing of elements
 */

/* Focused element */
.gvs-viewer .gvs-main
{
  position: relative;
  width: 100%;
  height: 100%;
  z-index: 0;
}

.gvs-viewer .gvs-main {
  display: flex;
  flex-direction: column;
}

/* Non-focused element */
.gvs-viewer .gvs-mini
{
  position: absolute;
  top: unset;
  bottom: 10px;
  left: 10px;
  height: 30%;
  min-height: 232px;
  aspect-ratio: 1 / 1;
  z-index: 1;
}

.gvs.gvs-viewer:not(.gvs-has-mini) .gvs-mini,
.gvs.gvs-viewer.gvs-mini-hidden .gvs-mini {
  display: none;
}

@container (max-width: 576px) {       /* Special rule for small containers */
  .gvs-viewer .gvs-mini {
    max-width: 166px;
    max-height: 110px;
    min-height: unset;
    width: 50%;
    height: 30%;
  }
}

.gvs-viewer .gvs-map.maplibregl-map {
	position: absolute;
	top: 0;
	right: 0;
	left: 0;
	bottom: 0;
}

/* PSV fulfilling its parent */
.gvs-viewer .gvs-psv,
.gvs-viewer .gvs-popup {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
}


/*
 * Styling
 */


/* Non-focused element */
.gvs-viewer .gvs-mini,
.gvs-viewer .gvs-mini .psv-container,
.gvs-viewer .gvs-mini .gvs-map
{
  border-radius: 10px;
}

/* PSV under widgets */
.gvs-viewer .psv-container {
  z-index: 0;
}

/* No PSV loader */
.gvs-viewer .psv-loader {
  display: none;
}

/* Overlay under navbar */
.gvs-viewer .psv-overlay {
  z-index: 89;
}

/* Popup */
.gvs-viewer .gvs-popup {
  display: flex;
  justify-content: center;
  align-items: center;
  transition: opacity 0.1s;
  z-index: 10;
  visibility: visible;
  opacity: 1;
}

.gvs-viewer .gvs-popup.gvs-hidden {
  display: flex !important;
  opacity: 0;
  visibility: hidden;
}

.gvs-viewer .gvs-popup-backdrop {
  position: absolute;
  background: rgba(0, 0, 0, 0.85);
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  z-index: 10;
}

.gvs-viewer .gvs-popup div.gvs-widget-bg {
  max-width: 90%;
  max-height: 90%;
	position: absolute;
	padding: 15px;
	z-index: 10;
	border-radius: 25px;
  overflow-y: auto;
}

.gvs-viewer #gvs-popup-btn-close {
  position: absolute;
  top: 15px;
  right: 15px;
  width: 24px;
  min-width: unset;
  height: 24px;
  border-radius: 12px;
}

.gvs-viewer .gvs-popup table {
  border-collapse: collapse;
  font-size: 0.9rem;
  width: 100%;
}

.gvs-viewer .gvs-popup thead {
  background-color: var(--blue-pale);
}

.gvs-viewer .gvs-popup th[scope="row"] {
  text-align: left;
}

.gvs-viewer .gvs-popup th, .gvs-popup td {
  border: 1px solid var(--grey-semi-dark);
  padding: 8px 10px;
  max-width: 600px;
}

.gvs-viewer .gvs-popup .gvs-table-light th[scope="row"] {
  width: 30%;
}

.gvs-viewer .gvs-popup .gvs-table-light th,
.gvs-viewer .gvs-popup .gvs-table-light td {
  border: none;
  padding: 5px 10px;
  text-align: left;
}

.gvs-viewer .gvs-popup .gvs-table-light .gvs-td-with-id {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.gvs-viewer .gvs-popup table:not(.gvs-table-light) td:last-of-type {
  text-align: center;
}

.gvs-viewer .gvs-popup table:not(.gvs-table-light) tbody > tr:nth-of-type(even) {
  background-color: var(--grey);
}

.gvs-viewer .gvs-popup details summary {
  font-size: 1.0em;
	line-height: 1.0em;
	font-weight: 500;
	margin: 15px 0;
  cursor: pointer;
}

.gvs-viewer .gvs-popup details summary svg {
	height: 18px;
	vertical-align: sub;
	margin-right: 2px;
}

.gvs-viewer .gvs-metadata-actions {
  justify-content: center;
  font-size: 0.9rem;
}