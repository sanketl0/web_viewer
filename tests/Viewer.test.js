import Viewer from "../src/Viewer";

jest.mock("@photo-sphere-viewer/core", () => ({
	Viewer: class {},
	SYSTEM: {},
	DEFAULTS: {},
}));

jest.mock("@photo-sphere-viewer/equirectangular-tiles-adapter", () => ({
	EquirectangularTilesAdapter: jest.fn(),
}));

jest.mock("@photo-sphere-viewer/virtual-tour-plugin", () => ({
	VirtualTourPlugin: jest.fn(),
}));

jest.mock("maplibre-gl", () => ({
	addProtocol: jest.fn(),
	Map: class {},
}));

global.AbortSignal = { timeout: jest.fn() };

const API_URL = "http://localhost:5000/api/search";


describe("constructor", () => {
	it("works", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		expect(v).toBeDefined();
	});
});

describe("destroy", () => {
	it("works", () => {
		const d = document.createElement("div");
		d.id = "geovisio";
		document.body.appendChild(d);
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure("geovisio");
		v._hash = { destroy: jest.fn() };
		v._widgets = { destroy: jest.fn() };
		v.map = { destroy: jest.fn() };
		v.psv = { destroy: jest.fn() };

		v.destroy();

		expect(v._api).toBeUndefined();
		expect(v._hash).toBeUndefined();
		expect(v.map).toBeUndefined();
		expect(v.psv).toBeUndefined();
		expect(d.innerHTML).toBe("");
	});
});

describe("_initContainerStructure", () => {
	beforeEach(() => {
		document.body.innerHTML = "";
	});

	it("works with string", () => {
		const d = document.createElement("div");
		d.id = "geovisio";
		document.body.appendChild(d);
		const v = new Viewer("geovisio", API_URL, { testing: true });
		v._initContainerStructure();
		expect([...d.children].find(n => n.className.includes("gvs-main"))).toBeDefined();
		expect([...d.children].find(n => n.className.includes("gvs-mini"))).toBeDefined();
		expect(d.className).toBe("gvs gvs-viewer");
	});

	it("works with Element", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure();
		expect([...d.children].find(n => n.className.includes("gvs-main"))).toBeDefined();
		expect([...d.children].find(n => n.className.includes("gvs-mini"))).toBeDefined();
		expect(d.className).toBe("gvs gvs-viewer");
	});
});

describe("setPopup", () => {
	it("opens", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure();
		v.psv = {
			startKeyboardControl: jest.fn(),
			stopKeyboardControl: jest.fn(),
		};
		v.setPopup(true, "BLABLABLA");
		expect(v.popupContainer.innerHTML).toMatchSnapshot();
		expect(v.popupContainer.classList.contains("gvs-hidden")).toBeFalsy();
	});

	it("closes", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { map: false, testing: true });
		v._initContainerStructure();
		v.psv = {
			autoSize: jest.fn(),
			startKeyboardControl: jest.fn(),
			stopKeyboardControl: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.setPopup(true, "BLABLABLA");
		v.setPopup(false);
		expect(v.popupContainer.classList.contains("gvs-hidden")).toBeTruthy();
	});

	it("reopens", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure();
		v.psv = {
			autoSize: jest.fn(),
			startKeyboardControl: jest.fn(),
			stopKeyboardControl: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.setPopup(true, "BLABLABLA");
		v.setPopup(false);
		v.setPopup(true);
		expect(v.popupContainer.innerHTML).toMatchSnapshot();
		expect(v.popupContainer.classList.contains("gvs-hidden")).toBeFalsy();
	});
});

describe("playSequence", () => {
	it("sends event", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		return expect(new Promise(resolve => {
			v.addEventListener("sequence-playing", resolve);
			v.playSequence();
		})).resolves.toBeDefined();
	});
});

describe("stopSequence", () => {
	it("sends event", async () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v.psv = {
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		}
		return expect(new Promise(resolve => {
			v.addEventListener("sequence-stopped", resolve);
			v.stopSequence();
		})).resolves.toBeDefined();
	});
});

describe("isSequencePlaying", () => {
	it("is true when sequence is playing", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v.psv = { addEventListener: jest.fn() };
		v.psv.goToNextPicture = jest.fn();
		v.playSequence();
		expect(v.isSequencePlaying()).toBe(true);
	});

	it("is false when sequence never played", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v.psv = { addEventListener: jest.fn() };
		expect(v.isSequencePlaying()).toBe(false);
	});

	it("is false when sequence stopped", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v.psv = {
			addEventListener: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.playSequence();
		v.stopSequence();
		expect(v.isSequencePlaying()).toBe(false);
	});
});

describe("isMapWide", () => {
	it("works with map small", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure(d);
		v.map = {};
		expect(v.isMapWide()).toBe(false);
	});

	it("works with map wide", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure(d);
		v.map = {
			keyboard: { enable: jest.fn() },
			resize: jest.fn(),
			getCanvas: () => ({ focus: jest.fn() })
		};
		v.psv = {
			autoSize: jest.fn(),
			stopKeyboardControl: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.setFocus("map");
		expect(v.isMapWide()).toBe(true);
	});

	it("fails if no map set", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure(d);
		expect(() => v.isMapWide()).toThrow(new Error("Map is not enabled"));
	});
});

describe("setFocus", () => {
	it("changes focus from map to pic", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure(d);
		v.dispatchEvent = jest.fn();
		v.map = {
			keyboard: { enable: jest.fn(), disable: jest.fn() },
			resize: jest.fn(),
			getCanvas: () => ({ focus: jest.fn() })
		};
		v.psv = {
			autoSize: jest.fn(),
			startKeyboardControl: jest.fn(),
			stopKeyboardControl: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.setFocus("map");
		expect(v.isMapWide()).toBe(true);
	});

	it("changes focus from pic to map", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure(d);
		v.dispatchEvent = jest.fn();
		v.map = {
			keyboard: { enable: jest.fn(), disable: jest.fn() },
			resize: jest.fn(),
			getCanvas: () => ({ focus: jest.fn() })
		};
		v.psv = {
			autoSize: jest.fn(),
			startKeyboardControl: jest.fn(),
			stopKeyboardControl: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.setFocus("pic");
		expect(v.isMapWide()).toBe(false);
	});

	it("skips event", async () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v._initContainerStructure(d);
		v.dispatchEvent = jest.fn();
		v.map = {
			keyboard: { enable: jest.fn(), disable: jest.fn() },
			resize: jest.fn(),
			getCanvas: () => ({ focus: jest.fn() })
		};
		v.psv = {
			autoSize: jest.fn(),
			startKeyboardControl: jest.fn(),
			stopKeyboardControl: jest.fn(),
			adapter: { __refresh: jest.fn() },
			needsUpdate: jest.fn(),
			_myVTour: { getCurrentNode: jest.fn() },
		};
		v.setFocus("map", true);
		expect(v.dispatchEvent.mock.calls).toEqual([]);
	});

	it("fails if param is invalid", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		expect(() => v.setFocus()).toThrow(new Error("Invalid focus value (should be pic or map)"));
	});

	it("fails if no map set", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { map: false, testing: true });
		expect(() => v.setFocus("map")).toThrow(new Error("Map is not enabled"));
	});
});

describe("getPicturesNavigation", () => {
	it("works", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { picturesNavigation: "pic", testing: true });
		expect(v.getPicturesNavigation()).toBe("pic");
		v._picNav = "seq";
		expect(v.getPicturesNavigation()).toBe("seq");
	});
});

describe("setPicturesNavigation", () => {
	it("works", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		const eventWatcher = jest.fn();
		v.addEventListener("pictures-navigation-changed", eventWatcher);
		expect(v.getPicturesNavigation()).toBe("any");
		v.setPicturesNavigation("pic");
		expect(v.getPicturesNavigation()).toBe("pic");
		v.setPicturesNavigation("seq");
		expect(v.getPicturesNavigation()).toBe("seq");
		v.setPicturesNavigation("any");
		expect(v.getPicturesNavigation()).toBe("any");
		expect(eventWatcher.mock.calls).toMatchSnapshot();
	});
});

describe("setFilters", () => {
	it("works", () => {
		const d = document.createElement("div");
		const v = new Viewer(d, API_URL, { testing: true });
		v.map = {
			_userLayers: new Set(["geovisio"]),
			getZoom: jest.fn(),
			reloadLayersStyles: jest.fn(),
			filterUserLayersContent: jest.fn(),
			getVisibleUsers: () => ["geovisio"],
			_hasGridStats: () => false,
		};
		v.dispatchEvent = jest.fn();
		Date.prototype.getDate = () => 8;
		Date.prototype.setDate = jest.fn();
		Date.prototype.toISOString = () => "2023-08-09T00:00:00Z";
		v.setFilters({
			"minDate": "2023-01-01",
			"maxDate": "2023-08-08",
			"camera": "sony",
			"type": "equirectangular",
			"theme": "age",
		});
		v.setFilters({});
		expect(v._mapTheme).toBe("age");
		expect(v.dispatchEvent.mock.calls).toMatchSnapshot();
		expect(v.map.reloadLayersStyles.mock.calls).toMatchSnapshot();
		expect(v.map.filterUserLayersContent.mock.calls).toMatchSnapshot();
	});
});