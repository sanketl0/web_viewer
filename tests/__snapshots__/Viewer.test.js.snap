// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setFilters works 1`] = `
Array [
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
]
`;

exports[`setFilters works 2`] = `
Array [
  Array [],
]
`;

exports[`setFilters works 3`] = `
Array [
  Array [
    "sequences",
    Array [
      "step",
      Array [
        "zoom",
      ],
      true,
      7,
      Array [
        "all",
        Array [
          ">=",
          Array [
            "get",
            "date",
          ],
          "2023-01-01",
        ],
        Array [
          "<=",
          Array [
            "get",
            "date",
          ],
          "2023-08-08",
        ],
        Array [
          "==",
          Array [
            "get",
            "type",
          ],
          "equirectangular",
        ],
        Array [
          ">=",
          Array [
            "get",
            "model",
          ],
          "sony                    ",
          Array [
            "collator",
            Object {
              "case-sensitive": false,
              "diacritic-sensitive": false,
            },
          ],
        ],
        Array [
          "<=",
          Array [
            "get",
            "model",
          ],
          "sonyzzzzzzzzzzzzzzzzzzzz",
          Array [
            "collator",
            Object {
              "case-sensitive": false,
              "diacritic-sensitive": false,
            },
          ],
        ],
      ],
    ],
  ],
  Array [
    "pictures",
    Array [
      "step",
      Array [
        "zoom",
      ],
      true,
      15,
      Array [
        "all",
        Array [
          ">=",
          Array [
            "get",
            "ts",
          ],
          "2023-01-01",
        ],
        Array [
          "<=",
          Array [
            "get",
            "ts",
          ],
          "2023-08-09",
        ],
        Array [
          "==",
          Array [
            "get",
            "type",
          ],
          "equirectangular",
        ],
        Array [
          ">=",
          Array [
            "get",
            "model",
          ],
          "sony                    ",
          Array [
            "collator",
            Object {
              "case-sensitive": false,
              "diacritic-sensitive": false,
            },
          ],
        ],
        Array [
          "<=",
          Array [
            "get",
            "model",
          ],
          "sonyzzzzzzzzzzzzzzzzzzzz",
          Array [
            "collator",
            Object {
              "case-sensitive": false,
              "diacritic-sensitive": false,
            },
          ],
        ],
      ],
    ],
  ],
  Array [
    "sequences",
    null,
  ],
  Array [
    "pictures",
    null,
  ],
]
`;

exports[`setPicturesNavigation works 1`] = `
Array [
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
]
`;

exports[`setPopup opens 1`] = `"<div class=\\"gvs-popup-backdrop\\"></div><div class=\\"gvs-widget-bg\\"><button id=\\"gvs-popup-btn-close\\" class=\\"gvs-btn gvs-widget-bg\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"xmark\\" class=\\"svg-inline--fa fa-xmark\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z\\"></path></svg></button>BLABLABLA</div>"`;

exports[`setPopup reopens 1`] = `"<div class=\\"gvs-popup-backdrop\\"></div><div class=\\"gvs-widget-bg\\"><button id=\\"gvs-popup-btn-close\\" class=\\"gvs-btn gvs-widget-bg\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"xmark\\" class=\\"svg-inline--fa fa-xmark\\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 384 512\\"><path fill=\\"currentColor\\" d=\\"M342.6 150.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L192 210.7 86.6 105.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L146.7 256 41.4 361.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0L192 301.3 297.4 406.6c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L237.3 256 342.6 150.6z\\"></path></svg></button>BLABLABLA</div>"`;
